common/net/credentials/alts/handshaker/handshaker.go

```diff
@@ -0,0 +1,408 @@
+)
+
+func init() {
+	for protocol, f := range altsRecordFuncs {
+		if err := conn.RegisterProtocol(protocol, f); err != nil {
+			panic(err)
+		}
+	}
+}
+
+func acquire() bool {
+	mu.Lock()
+	// If we need n to be configurable, we can pass it as an argument.
+	n := int64(1)
+	success := maxPendingHandshakes-concurrentHandshakes >= n
+	if success {
+		concurrentHandshakes += n
+	}
+	mu.Unlock()
+	return success
+}
+
+func release() {
+	mu.Lock()
+	// If we need n to be configurable, we can pass it as an argument.
+	n := int64(1)
+	concurrentHandshakes -= n
+	if concurrentHandshakes < 0 {
+		mu.Unlock()
+		panic("bad release")
+	}
+	mu.Unlock()
+}
+
+// data exchange request
+type kexExchangeRequest struct {
+	PubKey    []byte // the client's public key
+	PastelID  string
+	Signature []byte
+}
+
+// data exchange response
+type kexExchangeResponse struct {
+	PubKey    []byte // the server's public key
+	PastelID  string
+	Signature []byte
+}
+
+// ClientHandshakerOptions contains the client handshaker options that can
+// provided by the caller.
+type ClientHandshakerOptions struct {
+}
+
+// ServerHandshakerOptions contains the server handshaker options that can
+// provided by the caller.
+type ServerHandshakerOptions struct {
+}
+
+// DefaultClientHandshakerOptions returns the default client handshaker options.
+func DefaultClientHandshakerOptions() *ClientHandshakerOptions {
+	return &ClientHandshakerOptions{}
+}
+
+// DefaultServerHandshakerOptions returns the default client handshaker options.
+func DefaultServerHandshakerOptions() *ServerHandshakerOptions {
+	return &ServerHandshakerOptions{}
+}
+
+// altsHandshaker is used to complete a ALTS handshaking between client and server.
+type altsHandshaker struct {
+	// the record protocol
+	protocol string
+	// the record key
+	key []byte
+	// the connection to the peer.
+	conn net.Conn
+	// client handshake options.
+	clientOpts *ClientHandshakerOptions
+	// server handshake options.
+	serverOpts *ServerHandshakerOptions
+	// defines the side doing the handshake, client or server.
+	side alts.Side
+}
+
+// NewClientHandshaker creates a ALTS handshaker
+func NewClientHandshaker(_ context.Context, conn net.Conn, opts *ClientHandshakerOptions) (alts.Handshaker, error) {
+	return &altsHandshaker{
+		protocol:   recordxChaChaPolyProtocol,
+		conn:       conn,
+		clientOpts: opts,
+		side:       alts.ClientSide,
+	}, nil
+}
+
+// NewServerHandshaker creates a ALTS handshaker
+func NewServerHandshaker(_ context.Context, conn net.Conn, opts *ServerHandshakerOptions) (alts.Handshaker, error) {
+	return &altsHandshaker{
+		protocol:   recordxChaChaPolyProtocol,
+		conn:       conn,
+		serverOpts: opts,
+		side:       alts.ServerSide,
+	}, nil
+}
+
+// read and decode the handshake record
+func (s *altsHandshaker) readHandshake(value interface{}) error {
+	hdr := make([]byte, handshakeHeader)
+	// read handshake header from connection
+	if _, err := s.conn.Read(hdr); err != nil {
```

Comment:
Is there any guarantee that `s.conn.Read` will only return after reading exactly `len(hdr)` ? If no then a partial read can give a wrong body length and memory allocation will go wild. I see code like this in the network part of `kamedila` already and it makes supernode crash intermittently.

Reply:
you can give me where  use like this in`kamedila`

Reply:
https://github.com/pastelnetwork/gonode/blob/8a39b3090588b18708c48ee3b084d3981198f272/p2p/kademlia/message.go#L93
It should not be used like this.

Reply:
so, do you have any suggestion to replace ?